### PR TITLE
feat: add network wire byte

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
Added network wire byte. This is now independent of the network as byte.

Note: This method in favour of #6502.

Motivation and Context
---
Base nodes and wallets maintained numerous connections from which they could not sync or obtain useful information. This caused base nodes to try and follow other base nodes that were on a higher proof of work but on a different genesis block, and caused wallets to query base nodes that supplied information from a different blockchain. This byte can now be changed every time the genesis block changes.

How Has This Been Tested?
---
Passed all unit tests.

What process can a PR reviewer use to test or verify this change?
---

Breaking Changes
---

- [X] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

File: base_layer/p2p/src/lib.rs [hunk 1]
```diff
@@ -57,3 +57,44 @@ pub const MAJOR_NETWORK_VERSION: u8 = 0;
...
+    };
+    if !valid {
+        return Err(anyhow::anyhow!(
+            "Invalid network wire byte `{}` for network `{}`",
+            network_wire_byte,
+            network
+        ));
+    }
+    Ok(())
+}
```
Comment (sdbondi) [inline base_layer/p2p/src/lib.rs]: This should be a unit test rather - it's not necessary to check hard coded values at runtime - you can then make get_network_wire_byte infallible.

File: base_layer/p2p/src/lib.rs [hunk 2]
```diff
@@ -57,3 +57,44 @@ pub const MAJOR_NETWORK_VERSION: u8 = 0;
...
+    let network_wire_byte = match network {
+        Network::MainNet => 0,
+        Network::StageNet => 40,
+        Network::NextNet => 80,
+        Network::LocalNet => 120,
+        Network::Igor => 160,
+        Network::Esmeralda => 200,
+    };
+    verify_network_wire_byte_range(network_wire_byte, network)?;
+    Ok(network_wire_byte)
```
Comment (sdbondi) [inline base_layer/p2p/src/lib.rs]: How about making this a method on `Network` e.g. `network.as_wire_byte()`

File: common/src/configuration/network.rs [hunk 3]
```diff
@@ -86,6 +95,66 @@ impl Network {
...
+            // Choose a value in 'LOCAL_NET_RANGE' or assign 'self.as_byte()'
+            Network::LocalNet => self.as_byte(),
+            // Choose a value in 'IGOR_RANGE' or assign 'self.as_byte()'
+            Network::Igor => self.as_byte(),
+            // Choose a value in 'ESMERALDA_RANGE' or assign 'self.as_byte()'
+            Network::Esmeralda => self.as_byte(),
+        };
+        // The error is squashed to make this method infallible, and the method is called here to ensure that the
+        // range constants are used.
+        let _unused = self.verify_network_wire_byte_range(wire_byte);
```
Comment (sdbondi) [inline common/src/configuration/network.rs]: This does not need to happen at all - the `verify_network_wire_byte_range`  function can become a test that runs.

Comment (sdbondi) [reply common/src/configuration/network.rs]:
```rust
#[test]
fn check_constant_wire_bytes_are_in_correct_range() {
verify_network_wire_byte_range(Network::Esmeralda);
// etc..
}

 fn verify_network_wire_byte_range(self, network: Network) -> Result<(), String> {
        let network_wire_byte = network.as_wire_byte();
// 'LIVENESS_WIRE_MODE' is reserved for '0xa7'
        if network_wire_byte == LIVENESS_WIRE_MODE {
            panic!("Invalid network wire byte, cannot be '0x46', reserved for 'LIVENESS_WIRE_MODE'");
        }

        // Verify binned values
        let valid = match self {
            Network::MainNet => MAIN_NET_RANGE.contains(&network_wire_byte),
            Network::StageNet => STAGE_NET_RANGE.contains(&network_wire_byte),
            Network::NextNet => NEXT_NET_RANGE.contains(&network_wire_byte),
            Network::LocalNet => LOCAL_NET_RANGE.contains(&network_wire_byte),
            Network::Igor => IGOR_RANGE.contains(&network_wire_byte),
            Network::Esmeralda => ESMERALDA_RANGE.contains(&network_wire_byte),
        };
        if !valid {
          panic!(
                "Invalid network wire byte `{}` for network `{}`",
                network_wire_byte, network
            );
        }
    }

**Review by SWvheerden — APPROVED** [review]

**Pull request merged**